[WIP] Add support for generating ATen files during fbcode build

anderspapitto opened the pull request.

Yangqing commented:
Awesome work, thanks so much for looking into this - do we have an internal diff that checks whether this works?

anderspapitto commented:
yes, this PR is generated from the open-source person of an internal diff.

That said, this isn't quite ready to land at the moment

Commit:
Back out "Back out "Add support for generating ATen files during fbcode build""

Original commit changeset: 7b8de22d1613

I'm re-sending this diff exactly as it was approved and
committed. Fixes to support @mode/opt will be sent separately for ease
of review.

Commit:
Enable building //caffe2:torch with @mode/opt

In @mode/opt, python runs out of a PAR, which breaks a lot of
assumptions in the code about where templates/ folders live relative
to __file__. Rather than introduce hacks with parutil, I simply turn
template_path into a parameter for all the relevant functions and
thread it through from the top level.

ezyang committed:
Fix probable lint failure

ezyang commented:
It would be good for there to be an explanation about what the fbcode-side constraints are, so future people editing the files later don't undo changes which are seemingly equivalent. I guess most of the faffing about has to do with making `--template-paths` configurable?
